Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -80,6 +114,92 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::flattenDescendantLayersIfNecessary(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
+
+    // Traverse all children in depth first, post-order
+    for (auto* child : m_children) {
+        child->flattenDescendantLayersIfNecessary(options);
+
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            child->flatten(options);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Flattens descendant layers to into this planes z=0 2D root.
+    if (m_flattenedLayer)
+        m_flattenedLayer = nullptr;
+
+    // Reset transformations as this layer is root 2D
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    m_layerTransforms.combined = { };
+    m_layerTransforms.combinedForChildren = { };
+#if USE(COORDINATED_GRAPHICS)
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
+    m_layerTransforms.futureLocalTransform = { };
+    m_layerTransforms.futureCombined = { };
+    m_layerTransforms.futureCombinedForChildren = { };
+#endif
+    ComputeTransformData data;
+    if (m_state.maskLayer)
+        m_state.maskLayer->computeTransformsRecursive(data);
+    if (m_state.replicaLayer)
+        m_state.replicaLayer->computeTransformsRecursive(data);
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
+
+    Region layerRegion;
+    computeFlattenedRegion(layerRegion);
+
+    auto flattenedLayerRect = layerRegion.bounds();
+    auto flattenedLayer = WTF::makeUnique<TextureMapperFlattenedLayer>(flattenedLayerRect);
```

Comment:
Why don't use reuse m_flattenedLayer? a single layer is rendered multiple times.

Reply:
I misunderstand. Never mind.